test/extended_equality.py

```diff
@@ -0,0 +1,244 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Utility for checking equality of data of Qiskit Experiments class which doesn't
+officially implement the equality dunder method.
+"""
+
+import dataclasses
+from typing import Any, List, Union
+
+import numpy as np
+import uncertainties
+from lmfit import Model
+from multimethod import multimethod
+from qiskit_experiments.curve_analysis.curve_data import CurveFitResult
+from qiskit_experiments.data_processing import DataAction, DataProcessor
+from qiskit_experiments.database_service.utils import (
+    ThreadSafeList,
+    ThreadSafeOrderedDict,
+)
+from qiskit_experiments.framework import (
+    ExperimentData,
+    BaseExperiment,
+    BaseAnalysis,
+    AnalysisResult,
+)
+from qiskit_experiments.visualization import BaseDrawer
+
+
+@multimethod
+def is_equivalent(
+    data1: object,
+    data2: object,
+):
+    """Equality check finally falls into this function."""
+    if data1 is None and data2 is None:
+        return True
+    if dataclasses.is_dataclass(data1) and dataclasses.is_dataclass(data2):
+        # Avoid dataclass asdict. This copies dictionary but experiment config dataclass
+        # may include type information which cannot be copied.
+        return is_equivalent(
+            data1.__dict__,
+            data2.__dict__,
+        )
+    return data1 == data2
```

Comment:
This leads to some surprising edge cases. For example `is_equivalent(np.array([]),123)` returns `array([], dtype=bool)` instead of `True` or `False`. `is_equivalent([0], np.zeros(2))` returns `array([ True,  True])` but I think it should be `False`. Also, `np.int64(2)` is considered equivalent to 2, but I'm not sure if this is the behavior we want since the current qpy serialization doesn't support numpy types. Either way, this should be documented in the docstring.

Reply:
Good catch. I'm bit surprised we are using such fragile logic (indeed this comes from existing code). I added type check after comparison. It now returns `False` when the return value is not bool type. Regarding the integer and float comparison, I added `np.integer` and `np.floating` (see [numpy scalars](https://numpy.org/doc/1.17/reference/arrays.scalars.html#scalars) doc) to `_check_integer` (newly added) and `_check_floats` so that we can compare value after typecasting to python built-in types.